src/Microsoft.AspNet.OData/Extensions/HttpRequestMessageExtensions.cs

```diff
@@ -410,6 +412,57 @@ public static IEnumerable<IODataRoutingConvention> GetRoutingConventions(this Ht
 
             return request.GetRequestContainer().GetServices<IODataRoutingConvention>();
         }
+        /// <summary>
+        /// Checks whether the request is a POST targeted at a resource path ending in /$query.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="oDataPath">The OData path.</param>
+        /// <returns>true if the request path has $query segment.</returns>
+        internal static bool IsQueryRequest(this HttpRequestMessage request, string oDataPath)
+        {
+            return request.Method.Equals(HttpMethod.Post) && 
+                oDataPath?.TrimEnd('/').EndsWith('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase) == true;
+        }
+
+        /// <summary>
+        /// Transforms a POST request targeted at a resource path ending in $query into a GET request. 
+        /// The query options are parsed from the request body and appended to the request URL.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="requestContainerFactory">Request container factory.</param>
+        internal static void TransformQueryRequest(this HttpRequestMessage request, Func<IServiceProvider> requestContainerFactory)
+        {
+            if (requestContainerFactory == null)
+            {
+                throw Error.ArgumentNull("requestContainerFactory");
+            }
+
+            IServiceProvider requestContainer = requestContainerFactory();
+
+            // Fetch parsers available in the request container for parsing the query options in the request body
+            IEnumerable<IODataQueryOptionsParser> queryOptionsParsers = requestContainer.GetRequiredService<IEnumerable<IODataQueryOptionsParser>>();
+            IODataQueryOptionsParser queryOptionsParser = queryOptionsParsers.FirstOrDefault(d => d.CanParse(request));
+
+            string mediaType = request.Content.Headers.ContentType?.MediaType ?? string.Empty;
+
+            if (queryOptionsParser == null)
+            {
+                throw new ODataException(string.Format(
+                    CultureInfo.InvariantCulture,
+                    SRResources.CannotFindParserForRequestMediaType,
+                    mediaType));
+            }
+
+            Stream requestStream = request.Content.ReadAsStreamAsync().Result;
+            string queryString = queryOptionsParser.ParseAsync(requestStream).Result;
+
+            string requestPath = request.RequestUri.LocalPath;
+            requestPath = requestPath.Substring(0, requestPath.LastIndexOf('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase));
+
+            Uri requestUri = request.RequestUri;
+            request.RequestUri = new UriBuilder(requestUri.Scheme, requestUri.Host, requestUri.Port, requestPath, queryString).Uri;
```

Comment:
>queryString [](start = 114, length = 11)

Note that, according to spec (see [http://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#sec_PassingQueryOptionsintheRequestBody](http://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#sec_PassingQueryOptionsintheRequestBody))
   "Query options specified in the request body and query options specified in the request URL are processed together."
i.e., if query options exist in the original request, they should be merged with query options specified in the body.

Note that the spec also says (see [http://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#sec_SystemQueryOptions](http://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#sec_SystemQueryOptions)):
   "The same system query option, irrespective of casing or whether or not it is prefixed with a $, MUST NOT be specified more than once for any resource."
so we should return an error if the same query option is specified both in the body and in the request.

Reply:
@mikepizzo Fixed.
Just to note, currently, when a system query option is specified more than once, e.g., `/People?$orderby=Name&$orderby=Name`, we respond with a 404 - NotFound. It made sense to keep the logic here lean by just merging the options in the request body to those in the request url and have any repeated query options caught downstream the same place it happens currently.

Reply:
:thumbs up:

---
In reply to: [488693700](https://github.com/OData/WebApi/pull/2191#discussion_r488693700) [](ancestors = 488693700)